src/Tizen.Applications.ComponentBased/Tizen.Applications.ComponentBased.Common/ComponentBasedApplicationBase.cs

```diff
@@ -0,0 +1,208 @@
+        private Interop.CBApplication.CBAppLifecycleCallbacks _callbacks;
+        private bool _disposedValue = false;
+
+        /// <summary>
+        /// Initializes the ComponentBasedApplicationBase class.
+        /// </summary>
+        /// <param name="typeInfo">The component type information.
+        /// The key should be a class type of BaseComponent subclass.
+        /// The value should be a component id which is declared in tizen-manifest.xml.
+        /// </param>
+        /// <exception cref="OutOfMemoryException">Thrown when failed because of out of memory.</exception>
+        /// <since_tizen> 6 </since_tizen>
+        public ComponentBasedApplicationBase(IDictionary<Type, string> typeInfo)
+        {
+            _callbacks.OnInit = new Interop.CBApplication.CBAppInitCallback(OnInitNative);
+            _callbacks.OnFinished = new Interop.CBApplication.CBAppFiniCallback(OnFinishedNative);
+            _callbacks.OnRun = new Interop.CBApplication.CBAppRunCallback(OnRunNative);
+            _callbacks.OnExit = new Interop.CBApplication.CBAppExitCallback(OnExitNative);
+            _callbacks.OnCreate = new Interop.CBApplication.CBAppCreateCallback(OnCreateNative);
+            _callbacks.OnTerminate = new Interop.CBApplication.CBAppTerminateCallback(OnTerminateNative);
+
+            foreach (var component in typeInfo)
+            {
+                RegisterComponent(component.Key, component.Value);
+            }
+        }
+
+        /// <summary>
+        /// Registers a component.
+        /// </summary>
+        /// <param name="compType">Class type</param>
+        /// <param name="compId">Component ID</param>
+        /// <exception cref="InvalidOperationException">Thrown when component type is already added.</exception>
+        /// <since_tizen> 6 </since_tizen>
+        public void RegisterComponent(Type compType, string compId)
+        {
+            foreach (ComponentFactoryBase factory in _componentFactories)
+            {
+                if (compType.BaseType == factory._classType)
+                    throw new InvalidOperationException("Already exist type");
+            }
+
+            if (typeof(BaseComponent).IsAssignableFrom(compType))
+            {
+                Log.Info(LogTag, "Add base comp");
+                BaseComponent b = Activator.CreateInstance(compType) as BaseComponent;
+                _componentFactories.Add(new BaseFactory(compType, compId, b.ComponentType, this));
+            }
+            else if (typeof(FrameComponent).IsAssignableFrom(compType))
+            {
+                Log.Info(LogTag, "Add frame comp");
+                _componentFactories.Add(new FrameFactory(compType, compId, this));
+            }
+            else if (typeof(ServiceComponent).IsAssignableFrom(compType))
+            {
+                Log.Info(LogTag, "Add service comp");
+                _componentFactories.Add(new ServiceFactory(compType, compId, this));
+            }
```

Comment:
You should throw exception if comType is not subtype of `BaseComponent`

Reply:
Done.
I added else statement and throw an exception.